### PR TITLE
[TypeChecker] Treat tuples specially while validating checked casts

### PR DESCRIPTION
**Pull request by xedin**
Based on the checked cast behavior don't warn about unrelated
casts between tuples unless their sizes or labels differ.

Resolves: rdar://problem/56436235

### CODE REVIEW

**Comment by xedin** [comment]
@slavapestov Does this make sense? @weissi told me that "always fails" warning is affecting SwiftNIO.

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -4165,24 +4170,31 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
...
+  if (auto toTuple = toType->getAs<TupleType>()) {
+    if (auto fromTuple = fromType->getAs<TupleType>()) {
+      if (fromTuple->getNumElements() != toTuple->getNumElements())
         return failed();
+
+      for (unsigned i = 0, n = toTuple->getNumElements(); i != n; ++i) {
+        const auto &fromElt = fromTuple->getElement(i);
+        const auto &toElt = toTuple->getElement(i);
+
+        if (fromElt.getName() != toElt.getName())
```
Comment (CodaFi) [inline lib/Sema/TypeCheckConstraints.cpp]: Can you add a regression test that mismatches the labels?

**Review by CodaFi — APPROVED** [review]
Never mind my comments, I can't read today.

Typing rule LGTM.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**